### PR TITLE
#166789994 Users Should Be Able to Reset Password via Email

### PR DESCRIPTION
**Pull request by Hervera**
#### What does this PR do?
It lets a user to reset the password via email

#### Description of Task to be completed
- Add nodemailer
- write send mail functionalities
- write reset password functionalities
- Add tests

#### How should this be manually tested?
1. Run the following commands in Terminal
```
git clone git@github.com:andela/ah-92explorers-backend.git
cd ah-92explorers-backend
git checkout ft-reset-password-via-email-166789994
npm install
mv .env.sample .env
npm run start
```
2. Send the reset password link using email
- use `POST` method with your registered email in the body in this route `http://localhost:3000/api/v1/password`

3. Visit your email to get the reset password link which has a token

4. Reset Password
- use `PUT` method with the token received and add new password in the body using this route `http://localhost:3000/api/v1/password` 

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#166789994](https://www.pivotaltracker.com/story/show/166789994)

#### Screenshots (if appropriate)
Screen Shot 2019-07-09 at 12 42 37

Screen Shot 2019-07-05 at 14 00 34

Screen Shot 2019-07-09 at 12 46 05

Screen Shot 2019-07-05 at 14 03 07

Screen Shot 2019-07-09 at 12 47 12

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I don't have more than 3 major commits on this PR

### CODE REVIEW

File: helpers/sendEmail.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import Mailer from './mailer';
+
+const sendEmail = (userEmail, token, template) => new Promise(async (resolve, reject) => {
```
Comment (Hervera) [inline helpers/sendEmail.js]: Use `async/await` instead of `new Promise()`

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]
Please rebase from develop before further reviewing

File: controllers/resetPasswordController.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import jwt from 'jsonwebtoken';
+import bcrypt from 'bcryptjs';
+import dotenv from 'dotenv';
+import models from '../models/index';
+import sendEmail from '../helpers/sendEmail';
+
+
+dotenv.config();
+const User = models.users;
```
Comment (kodek-sleuth) [inline controllers/resetPasswordController.js]: Please de-structure this line and please use a small cased User object

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]

File: index.js [hunk 3]
```diff
@@ -25,6 +25,11 @@ app.get('/swagger.json', (req, res) => {
   res.setHeader('Content-Type', 'application/json');
   res.send(swagger.swaggerSpec);
 });
+
+app.get('/reset-password/:token', (req, res) => {
```
Comment (elemanhillary-zz) [inline index.js]: line `29` of `index.js` should be put in `routers` folder, kindly follow the project structure to fix that.

**Review by elemanhillary-zz — COMMENTED** [review]
fix the issues reflected in the `review comments`

File: index.js (hunk 3 above)
Comment (elemanhillary-zz) [inline index.js]: line `29` of `index.js` should be in `routes` folder, kindly check the project structure to fix it

**Review by elemanhillary-zz — CHANGES_REQUESTED** [review]

**Review by elemanhillary-zz — CHANGES_REQUESTED** [review]
fix merge conflicts

**Comment by Hervera** [comment]
> fix merge conflicts

**Review by niyoceles — DISMISSED** [review]

**Review by IsaiahRn — DISMISSED** [review]
It is ready to be merged

**Review by elemanhillary-zz — DISMISSED** [review]

**Review by kodek-sleuth — DISMISSED** [review]
This PR is ready too be merged

**Review by Krasivaya — APPROVED** [review]

**Review by Krasivaya — APPROVED** [review]

**Review by Krasivaya — DISMISSED** [review]
Ready to be merged

**Review by niyoceles — DISMISSED** [review]

**Review by elemanhillary-zz — COMMENTED** [review]
ready for merge

**Review by elemanhillary-zz — CHANGES_REQUESTED** [review]
ready for merge

**Review by elemanhillary-zz — DISMISSED** [review]
ready for merge

**Review by kodek-sleuth — DISMISSED** [review]

**Review by niyoceles — DISMISSED** [review]

File: .env.example [hunk 4]
```diff
@@ -1,3 +1,8 @@
-DATABASE_URL=postgres://wmqxaqjr:tTHUvMlywe2owxdrqEaofzdC2RKF4XYp@raja.db.elephantsql.com:5432/wmqxaqjr
-DATABASE_TEST=postgres://wmqxaqjr:tTHUvMlywe2owxdrqEaofzdC2RKF4XYp@raja.db.elephantsql.com:5432/wmqxaqjr
-SECRET=qwertyuiop[]\';lkjhgfdsa`zxcvbnm,./+_)(*&^%$#@!)
+DATABASE_URL=postgresql://dbusername:dbpassword@localhost:5432/dbname
+DATABASE_TEST=postgresql://dbusername:dbpassword@localhost:5432/dbnamefortest
+
+SENDER_EMAIL=
```
Comment (kodek-sleuth) [inline .env.example:4]: Please complete these lines

File: controllers/resetPasswordController.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        };
+        const token = jwt.sign(payload, SECRET, expirationTime);
+        req.body.token = token;
+        req.body.template = 'resetPassword';
+        sendEmail(user.email, token, 'resetPassword');
+        return res.status(200).send({ message: 'We have e-mailed a password reset link, Check your email!' });
+      }
+      return res.status(404).json({ error: 'The email provided does not exist' });
+    } catch (error) {
+      return res.status(500).json({ error });
```
Comment (kodek-sleuth) [inline controllers/resetPasswordController.js]: its not advisable to return an error object in that way, its better to specify the error you want returned

**Review by kodek-sleuth — CHANGES_REQUESTED** [review]

**Review by elemanhillary-zz — DISMISSED** [review]
ready for merge

**Review by niyoceles — DISMISSED** [review]

**Review by elemanhillary-zz — DISMISSED** [review]
ready for merge

**Review by elemanhillary-zz — APPROVED** [review]
ready for merge

**Review by kodek-sleuth — APPROVED** [review]

**Review by IsaiahRn — APPROVED** [review]
it can be approved

**Pull request merged**